### ISSUE TITLE
URI validation can take a very long time

### ISSUE THREAD

**Issue by yourpalal** [davishmcclurg/json_schemer#46]
This could potentially lead to a DOS attack, if you are validating user-supplied data.

Example timing of validating similar urls. The key pattern that I have found seems to be long parameters in a URL that ends with a space. There may be other ways to achieve the same effect.

```
checking 5 -> http://example.com/?zzzzz=uhoh 
	checked in 0 seconds
checking 10 -> http://example.com/?zzzzzzzzzz=uhoh 
	checked in 0 seconds
checking 15 -> http://example.com/?zzzzzzzzzzzzzzz=uhoh 
	checked in 1 seconds
checking 20 -> http://example.com/?zzzzzzzzzzzzzzzzzzzz=uhoh 
	checked in 23 seconds
checking 25 -> http://example.com/?zzzzzzzzzzzzzzzzzzzzzzzzz=uhoh 
```
the version with 25 `z`'s takes a very long time, I haven't let it finish yet.

The above output was made with the following script:

```ruby
schema = JSONSchemer.schema({
                              "type" =>  "object",
                              "properties" =>  {
                                "url" =>  {
                                  "type" =>  "string",
                                  "format" =>  "uri"
                                }
                              },
                              "required" =>  %w(url)
                            })

(1..10).each do |i|
  url = "http://example.com/?#{'z' * i * 5}=uhoh "

  puts "checking #{i * 5} -> #{url}"

  start = DateTime.now
  schema.validate({"url" => url}).to_a
  done = DateTime.now

  puts "\tchecked in #{done.to_i - start.to_i} seconds"
end
```

I think perhaps switching to using [URI::Parser](https://ruby-doc.org/stdlib-2.6.5/libdoc/uri/rdoc/URI.html#Parser) could be a way to fix this problem. I'm not sure how the performance of the library would be affected by such a change, but it can parse all of the URIs in the above example very quickly.

**Comment by yourpalal** [comment]
note: deleting the space at the end of the URL fixes this issue

**Comment by davishmcclurg** [comment]
Thanks for the report! I was able to get it working using `URI::RFC3986_PARSER` like you suggested: https://github.com/davishmcclurg/json_schemer/pull/50

Great to get rid of those terrifying regular expressions. 🍕 

Released in 0.2.8.

**Comment by yourpalal** [comment]
Awesome!! Thanks @davishmcclurg  :) Also, thanks more generally for the work on the gem!
